Sources/SwiftProtobuf/TextFormatScanner.swift

```diff
@@ -357,26 +272,153 @@ internal struct TextFormatScanner {
-                p += 1
-                skipWhitespace()
-                return s
+    /// Scan a string that encodes a byte field, return a count of
+    /// the number of bytes that should be decoded from it
+    private mutating func validateAndCountBytesFromString(terminator: UInt8) throws -> Int {
+      var count = 0
+      let start = p
+      while p != end {
+        let byte = p[0]
+        p += 1
+        if byte == terminator {
+          p = start
+          return count
+        }
+        switch byte {
+        case asciiBackslash: //  "\\"
+          if p != end {
+            let escaped = p[0]
+            p += 1
+            switch escaped {
+              case asciiZero...asciiSeven: // '0'...'7'
+                // C standard allows 1, 2, or 3 octal digits.
+                if p != end, p[0] >= asciiZero, p[0] <= asciiSeven {
+                  p += 1
+                  if p != end, p[0] >= asciiZero, p[0] <= asciiSeven {
+                    if escaped > asciiThree {
+                       // Out of range octal: three digits and first digit is greater than 3
+                      throw TextFormatDecodingError.malformedText
+                    }
+                    p += 1
+                  }
+                }
+                count += 1
+              case asciiLowerX: // 'x' hexadecimal escape
+                // C standard allows any number of digits after \x
+                // We ignore all but the last two
+                var hexDigits = 0
+                while p != end && fromHexDigit(p[0]) != nil {
+                  hexDigits += 1
+                  p += 1
+                }
+                if hexDigits == 0 {
+                  throw TextFormatDecodingError.malformedText // Hex escape must have at least 1 digit
+                }
+                count += 1
+              case asciiLowerA, // \a ("alert")
+                   asciiLowerB, // \b
+                   asciiLowerF, // \f
+                   asciiLowerN, // \n
+                   asciiLowerR, // \r
+                   asciiLowerT, // \t
+                   asciiLowerV, // \v
+                   asciiSingleQuote, // \'
+                   asciiDoubleQuote, // \"
+                   asciiQuestionMark, // \?
+                   asciiBackslash: // \\
+                count += 1
+              default:
+                throw TextFormatDecodingError.malformedText // Unrecognized escape
             }
+          }
+        default:
+          count += 1
+        }
+      }
+      throw TextFormatDecodingError.malformedText
+    }
+
+    /// Protobuf Text format uses C ASCII conventions for
+    /// encoding byte sequences, including the use of octal
+    /// and hexadecimal escapes.
+    ///
+    /// Assumes that validateAndCountBytesFromString() has already
+    /// verified the correctness.  So we get to avoid error checks here.
+    private mutating func parseBytesFromString(terminator: UInt8, into data: inout Data) {
```

Comment:
The count is the number of bytes after decoding.  I could add a separate count for the number of bytes before decoding, but I don't think it would actually simplify things.  Using the same terminator checks means that the validation and parsing functions have fundamentally identical structures.